File: EMF/java/edgex-emf/src/main/java/org/edgexfoundry/emf/EMFSubscriber.java
package org.edgexfoundry.emf;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.locks.ReentrantLock;

import org.edgexfoundry.domain.core.Event;
import org.edgexfoundry.emf.protobufevent.EMFEventConverter;
import org.edgexfoundry.support.logging.client.EdgeXLogger;
import org.edgexfoundry.support.logging.client.EdgeXLoggerFactory;
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Poller;

/**
 * This class represents the EMF Subscriber.
 */
public class EMFSubscriber {

    private String mServiceName;
    private EMFSubCallback mCallback;
    private String mIp;
    private int mPort;
    private Thread mThread;

    // ZMQ Subscriber socket
    private ZMQ.Socket mSubscriber;
    private ZMQ.Context mContext;

    // ZMQ poller
    private Poller mPoller;

    // ZMQ shut-down socket
    private ZMQ.Socket mShutdownServer;
    private ZMQ.Socket mShutdownClient;

    // Thread safety lock
    private ReentrantLock mSubLock;

    private final String TCP_PREFIX = "tcp://";
    private final String INPROC_PREFIX = "inproc://shutdown-";
    private final String TOPIC_PATTERN = "[a-zA-Z0-9-_./]+";
    private final static EdgeXLogger logger = EdgeXLoggerFactory
            .getEdgeXLogger(EMFSubscriber.class);

    /**
     * Subscribe to publisher at given IP and Port.
     *
     * @param ip
     *            IP address
     * @param port
     *            port number
     * @param msgType
     *            Message type of the message/event to be received from
     *            publisher. {@link EMFMessageType}
     * @param callback
     *            {@link EMFCallback}
     */
    public EMFSubscriber(String ip, int port, EMFSubCallback callback) {
        mIp = ip;
        mPort = port;
        mCallback = callback;
        mContext = EMFAPI.getInstance().getContext();
        mSubLock = new ReentrantLock(true);
    }

    /**
     * Subscribe to given serviceName.
     *
     * @param serviceName
     *            service for which subscription is required.
     * @param callback
     *            {@link EMFSubCallback}
     */
    public EMFSubscriber(String serviceName, EMFSubCallback callback) {
        mServiceName = serviceName;
        // For given serviceName get the ip and port [TODO]
        mCallback = callback;
        mContext = EMFAPI.getInstance().getContext();
        mSubLock = new ReentrantLock(true);
    }

    // finalize method to be called by Java Garbage collector before destroying
    // this object.
    @Override
    protected void finalize() {
        logger.debug("In finalize");
        stop();
    }

    /**
     * Interface to receive message callback from EMF subscriber.
     */
    public interface EMFSubCallback {
        /**
         * Invoked when message is received.
         *
         * @param event
         *            {@link Event}
         */
        public void onMessageCB(Event event);

        /**
         * Invoked when message is received for a specific topic.
         *
         * @param event
         *            {@link Event}
         */
        public void onMessageCB(String topic, Event event);
    }

    /**
     * Create filter as data consumer.
     *
     * @param filter
     *            {@link EMFFilter}
     *
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode createFilter(EMFFilter filter) {
        return EMFErrorCode.EMF_OK;
    }

    /**
     * Starts SUB instance.
     *
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode start() {
        if (null == mContext) {
            logger.error("Context is null");
            return EMFErrorCode.EMF_ERROR;
        }

        String address = getInProcUniqueAddress();
        try {
            mSubLock.lock();
            // Shutdown server sockets
            if (null == mShutdownServer) {
                mShutdownServer = mContext.socket(ZMQ.PAIR);
                mShutdownServer.bind(address);
            }

            // Shutdown client sockets
            if (null == mShutdownClient) {
                mShutdownClient = mContext.socket(ZMQ.PAIR);
                mShutdownClient.connect(address);
            }

            // Subscriber socket
            if (null == mSubscriber) {
                mSubscriber = mContext.socket(ZMQ.SUB);
                mSubscriber.connect(getSocketAddress());
            }

            // Register sockets to poller
            if (null == mPoller) {
                mPoller = mContext.poller(2);
                mPoller.register(mSubscriber);
                mPoller.register(mShutdownClient);
            }
        } catch (Exception e) {
            logger.error("Exception while starting subscriber: " + e.getMessage());
            mSubLock.unlock();
            stop();
            return EMFErrorCode.EMF_ERROR;
        }

        // Receiver thread
        if (null == mThread) {
            mThread = new Thread(new Runnable() {
                public void run() {
                    receive();
                }
            });
            mThread.start();
            logger.debug("Receiver thread started");
        }

        mSubLock.unlock();
        return EMFErrorCode.EMF_OK;
    }

    private String getSocketAddress() {
        return TCP_PREFIX + mIp + ":" + mPort;
    }

    private String getInProcUniqueAddress() {
        String address = INPROC_PREFIX + UUID.randomUUID().toString();
        return address;
    }

    /**
     * Subscribe for event/messages.
     *
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode subscribe() {
        return subscribeInternal("");
    }

    /**
     * Subscribe for event/messages on a particular topic.
     *
     * @param topic
     *            Topic to be subscribed.
     * @return {@link EMFErrorCode}
     *
     * @note (1) Topic name should be as path format. For example:
     *       home/livingroom/ (2) Topic name can have letters [a-z, A-z],
     *       numerics [0-9] and special characters _ - . and /
     */
    public EMFErrorCode subscribe(String topic) {
        // validate the topic
        String validTopic = validateTopic(topic);
        if (null == validTopic) {
            logger.error("Invalid topic: " + topic);
            return EMFErrorCode.EMF_INVALID_TOPIC;
        }
        logger.debug("Topic is: " + validTopic);
        return subscribeInternal(validTopic);
    }

    private EMFErrorCode subscribeInternal(String topic) {
        if (null == mSubscriber) {
            logger.error("Subscriber is null");
            return EMFErrorCode.EMF_ERROR;
        }

        // subscribe for messages
        try {
            mSubLock.lock();
            if (null != mSubscriber) {
                mSubscriber.subscribe(topic.getBytes());
            } else {
                return EMFErrorCode.EMF_ERROR;
            }
        } finally {
            mSubLock.unlock();
        }
        logger.debug("subscribed for events");
        return EMFErrorCode.EMF_OK;
    }

    /**
     * Subscribe for event/messages on given list of topics. On any of the topic
     * in list, if it failed to subscribe events it will return
     * EMF_ERROR/EMF_INVALID_TOPIC.
     *
     * @param topic
     *            List of topics to be subscribed.
     * @return {@link EMFErrorCode}
     *
     * @note (1) Topic name should be as path format. For example:
     *       home/livingroom/ (2) Topic name can have letters [a-z, A-z],
     *       numerics [0-9] and special characters _ - . and /
     */
    public EMFErrorCode subscribe(List<String> topics) {
        if (null == topics) {
            logger.error("Topic list is null");
            return EMFErrorCode.EMF_INVALID_TOPIC;
        }

        EMFErrorCode result = EMFErrorCode.EMF_OK;
        for (String topic : topics) {
            result = subscribe(topic);
            if (result != EMFErrorCode.EMF_OK) {
                return result;
            }
        }
        return result;
    }

    private void receive() {
        byte[] data = null;
        byte[] topicBytes = null;
        Event event = null;

        while (null != mThread && !mThread.isInterrupted()) {
            if (null == mSubscriber || null == mPoller) {
                logger.error("Subscriber or poller is null");
                return;
            }

            mPoller.poll();
            if (mPoller.pollin(0)) {
                try {
                    mSubLock.lock();
                    if (null != mSubscriber) {
                        data = mSubscriber.recv();
                        if (mSubscriber.hasReceiveMore()) {
                            topicBytes = data;
                            data = mSubscriber.recv();
                        }
                    }
                } catch (Exception e) {
                    logger.error("Exception while receiving: " + e.getMessage());
                } finally {
                    mSubLock.unlock();
                }

                logger.debug("Event received");
                event = EMFEventConverter.toEdgeXEvent(data);
                if (null != event) {
                    if (null == topicBytes) {
                        mCallback.onMessageCB(event);
                    } else {
                        String recvTopic = new String(topicBytes);
                        logger.debug("Topic: " + recvTopic);
                        mCallback.onMessageCB(recvTopic, event);
                    }
                }
            } else if (mPoller.pollin(1)) {
                logger.debug("Received shut down request");
                break;
            }

            data = null;
            topicBytes = null;
        }
    }

    /**
     * Un-subscribe for all the events from publisher.
     *
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode unSubscribe() {
        return unSubscribeInternal("");
    }

    /**
     * Un-subscribe for a specific topic events.
     *
     * @param topic
     *            topic to be unsubscribed.
     * @return {@link EMFErrorCode}
     *
     * @note (1) Topic name should be as path format. For example:
     *       home/livingroom/ (2) Topic name can have letters [a-z, A-z],
     *       numerics [0-9] and special characters _ - . and /
     */
    public EMFErrorCode unSubscribe(String topic) {
        // validate topic
        String validTopic = validateTopic(topic);
        if (null == validTopic) {
            logger.error("Invalid topic: " + topic);
            return EMFErrorCode.EMF_INVALID_TOPIC;
        }
        logger.debug("Topic is: " + validTopic);
        return unSubscribeInternal(topic);

    }

    private EMFErrorCode unSubscribeInternal(String topic) {
        if (null == mSubscriber) {
            logger.error("Subscriber is null");
            return EMFErrorCode.EMF_ERROR;
        }

        try {
            mSubLock.lock();
            if (null != mSubscriber) {
                mSubscriber.unsubscribe(topic.getBytes());
            } else {
                return EMFErrorCode.EMF_ERROR;
            }
        } finally {
            mSubLock.unlock();
        }
        logger.debug("Un-subscribed for events");
        return EMFErrorCode.EMF_OK;
    }

    /**
     * Un-subscribe for given list of topic events. On any of the topic in list,
     * if it failed to unsubscribe events it will return
     * EMF_ERROR/EMF_INVALID_TOPIC.
     *
     * @param topics
     *            List of topics to be unsubscribed.
     * @return {@link EMFErrorCode}
     *
     * @note (1) Topic name should be as path format. For example:
     *       home/livingroom/ (2) Topic name can have letters [a-z, A-z],
     *       numerics [0-9] and special characters _ - . and /
     */
    public EMFErrorCode unSubscribe(List<String> topics) {
        if (null == topics) {
            logger.error("Topic list is null");
            return EMFErrorCode.EMF_INVALID_TOPIC;
        }
        EMFErrorCode result = EMFErrorCode.EMF_OK;
        for (String topic : topics) {
            result = unSubscribe(topic);
            if (result != EMFErrorCode.EMF_OK) {
                return result;
            }
        }
        return result;
    }

    /**
     * Stop the SUB instance.
     *
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode stop() {

        try {
            mSubLock.lock();

            // Send a shutdown message to receiver thread
            if (null != mShutdownServer) {
                boolean result = mShutdownServer.send("shutdown");
                logger.debug("Shutdown send result: " + result);
            }

            // wait for receiver thread to stop
            try {
                if (null != mThread) {
                    mThread.join();
                }
            } catch (InterruptedException e) {
                logger.error("Thread join exception" + e.getMessage());
            }

            // Unregister sockets from poller
            if (null != mPoller) {
                mPoller.unregister(mSubscriber);
                mPoller.unregister(mShutdownClient);
            }

            // close shut down sockets
            if (null != mShutdownClient) {
                mShutdownClient.close();
            }
            if (null != mShutdownServer) {
                mShutdownServer.close();
            }

            // Close subscriber
            if (null != mSubscriber) {
                mSubscriber.close();
            }

            mSubscriber = null;
            mPoller = null;
            mShutdownClient = null;
            mShutdownServer = null;
            mThread = null;
        } catch (Exception e) {
            logger.error("Exception while stopping subscriber: " + e.getMessage());
            return EMFErrorCode.EMF_ERROR;
        } finally {
            mSubLock.unlock();
        }

        logger.debug("Subscriber stopped");
        return EMFErrorCode.EMF_OK;
    }

    /**
     * Get the IP address.
     *
     * @return IP address as String.
     */
    public String getIp() {
        return mIp;
    }

    /**
     * Get the port of the subscriber.
     *
     * @return port number as integer.
     */
    public int getPort() {
        return mPort;
    }

    /**
     * Get the service name.
     *
     * @return Service name.
     */
    public String getServiceName() {
        return mServiceName;
    }

    private String validateTopic(String topic) {
        if (null == topic || topic.isEmpty()) {
            return null;
        }

        // check whether topic contains only alphabet, digits and special
        // characters _ - . and /
        if (!topic.matches(TOPIC_PATTERN)) {
            return null;
        }

        // check whether last character is forward slash or not
        // if not append and return the string
        if (!topic.endsWith("/")) {
            topic = topic + "/";
        }
        return topic;
    }
}

File: EMF/go/emf/emfsubscriber.go
package emf

import (
	proto "github.com/golang/protobuf/proto"
	zmq "github.com/pebbe/zmq4"
	"go.uber.org/zap"

	List "container/list"
	"strconv"
)

// Address prefix to bind subscriber.
const SUB_TCP_PREFIX = "tcp://"

// Callback to get all the subscribed events.
type EMFSubCB func(event Event)

// Callback to get all the subscribed events for a specific topic.
type EMFSubTopicCB func(topic string, event Event)

// Structure represents EMFSubscriber.
type EMFSubscriber struct {
	ip               string
	port             int
	subCallback      EMFSubCB
	subTopicCallback EMFSubTopicCB

	subscriber *zmq.Socket
	context    *zmq.Context

	isReceiverStarted bool
}

// Contructs EMFSubscriber.
func GetEMFSubscriber(ip string, port int, subCallback EMFSubCB, subTopicCallback EMFSubTopicCB) *EMFSubscriber {
	var instance *EMFSubscriber
	instance = &EMFSubscriber{}
	instance.ip = ip
	instance.port = port
	instance.subCallback = subCallback
	instance.subTopicCallback = subTopicCallback
	instance.context = GetInstance().getContext()
	InitLogger()
	if nil == instance.context {
		logger.Error("Context is null")
	}
	instance.subscriber = nil
	instance.isReceiverStarted = false
	return instance
}

func receive(subInstance *EMFSubscriber) {
	var data []byte
	var event Event
	var err error
	var more bool
	var topic string

	for subInstance.isReceiverStarted {
		if nil == subInstance.subscriber {
			logger.Error("subscriber or poller is null")
			break
		}
		data, err = subInstance.subscriber.RecvBytes(0)
		if err != nil {
			break
		}
		more, err = subInstance.subscriber.GetRcvmore()
		if err != nil {
			break
		}
		if more {
			topic = string(data[:])
			data, err = subInstance.subscriber.RecvBytes(0)
		}

		//change byte array to Event
		err := proto.Unmarshal(data, &event)
		if nil != err {
			logger.Error("Error in unmarshalling data")
		}

		if more {
			subInstance.subTopicCallback(topic, event)
		} else {
			subInstance.subCallback(event)
		}
	}
	logger.Debug("Received the shut down request")
}

// Starts SUB instance.
func (subInstance *EMFSubscriber) Start() EMFErrorCode {
	if nil == subInstance.context {
		logger.Error("Context is null")
		return EMF_ERROR
	}

	if nil == subInstance.subscriber {
		var err error
		subInstance.subscriber, err = zmq.NewSocket(zmq.SUB)
		if nil != err {
			logger.Error("Subscriber Socket creation failed")
			return EMF_ERROR
		}
		var address string = getSubSocketAddress(subInstance.ip, subInstance.port)
		err = subInstance.subscriber.Connect(address)
		if nil != err {
			logger.Error("Subscriber Socket connect failed")
			return EMF_ERROR
		}
		logger.Debug("Starting subscriber", zap.String("Address", address))
	}

	//call a go routine [new thread] for receiver
	if false == subInstance.isReceiverStarted {
		subInstance.isReceiverStarted = true
		go receive(subInstance)
	}
	return EMF_OK
}

func (subInstance *EMFSubscriber) subscribeInternal(topic string) EMFErrorCode {
	if nil != subInstance.subscriber {
		err := subInstance.subscriber.SetSubscribe(topic)
		if nil != err {
			logger.Error("subscribeInternal error occured")
			return EMF_ERROR
		}
	} else {
		logger.Error("subscriber is null")
		return EMF_ERROR
	}
	logger.Debug("subscribed for events")
	return EMF_OK
}

// Subscribe for event/messages.
func (subInstance *EMFSubscriber) Subscribe() EMFErrorCode {
	return subInstance.subscribeInternal("")
}

// Subscribe for event/messages on a particular topic.
func (subInstance *EMFSubscriber) SubscribeForTopic(topic string) EMFErrorCode {
	//validate the topic
	validTopic := sanitizeTopic(topic)
	if validTopic == "" {
		return EMF_INVALID_TOPIC
	}
	logger.Debug("subscribing for events", zap.String("Topic", validTopic))
	return subInstance.subscribeInternal(validTopic)
}

// Subscribe for event/messages on given list of topics. On any of the topic
// in list, if it failed to subscribe events it will return
// EMF_ERROR/EMF_INVALID_TOPIC.
//
// Note:
// (1) Topic name should be as path format. For example:home/livingroom/
// (2) Topic name can have letters [a-z, A-z], numerics [0-9] and special characters _ - / and .
func (subInstance *EMFSubscriber) SubscribeForTopicList(topicList List.List) EMFErrorCode {
	if topicList.Len() == 0 {
		return EMF_INVALID_TOPIC
	}
	for topic := topicList.Front(); topic != nil; topic = topic.Next() {
		result := subInstance.SubscribeForTopic(topic.Value.(string))
		if result != EMF_OK {
			return result
		}
	}
	return EMF_OK
}

func (subInstance *EMFSubscriber) unSubscribeInternal(topic string) EMFErrorCode {
	if nil != subInstance.subscriber {
		err := subInstance.subscriber.SetUnsubscribe(topic)
		if nil != err {
			logger.Error("subscriber is null")
			return EMF_ERROR
		}
	} else {
		return EMF_ERROR
	}
	return EMF_OK
}

// Un-subscribe all the events from publisher.
func (subInstance *EMFSubscriber) UnSubscribe() EMFErrorCode {
	return subInstance.unSubscribeInternal("")
}

// Un-subscribe specific topic events.
//
// Note:
// (1) Topic name should be as path format. For example:home/livingroom/
// (2) Topic name can have letters [a-z, A-z], numerics [0-9] and special characters _ - / and .
func (subInstance *EMFSubscriber) UnSubscribeForTopic(topic string) EMFErrorCode {
	//validate the topic
	validTopic := sanitizeTopic(topic)
	if validTopic == "" {
		return EMF_INVALID_TOPIC
	}
	logger.Debug("Unsubscribe for events", zap.String("Topic", validTopic))
	return subInstance.unSubscribeInternal(validTopic)
}

// Un-subscribe event/messages on given list of topics. On any of the topic
// in list, if it failed to unsubscribe events it will return
// EMF_ERROR/EMF_INVALID_TOPIC.
//
// Note:
// (1) Topic name should be as path format. For example:home/livingroom/ .
// (2) Topic name can have letters [a-z, A-z], numerics [0-9] and special characters _ - / and .
func (subInstance *EMFSubscriber) UnSubscribeForTopicList(topicList List.List) EMFErrorCode {
	if topicList.Len() == 0 {
		return EMF_INVALID_TOPIC
	}
	for topic := topicList.Front(); topic != nil; topic = topic.Next() {
		result := subInstance.UnSubscribeForTopic(topic.Value.(string))
		if result != EMF_OK {
			return result
		}
	}
	return EMF_OK
}

// Stops SUB instance.
func (subInstance *EMFSubscriber) Stop() EMFErrorCode {
	if nil != subInstance.subscriber {
		err := subInstance.subscriber.Close()
		if nil != err {
			logger.Error("Error while stopping subscriber")
			return EMF_ERROR
		}
	}
	subInstance.subscriber = nil
	subInstance.isReceiverStarted = false
	logger.Debug("Subscriber stopped")
	return EMF_OK
}

// Get Ip of publisher to which subscribed.
func (subInstance *EMFSubscriber) GetIP() string {
	return subInstance.ip
}

// Get Port of publisher to which subscribed.
func (subInstance *EMFSubscriber) GetPort() int {
	return subInstance.port
}

func getSubSocketAddress(ip string, port int) string {
	return string(SUB_TCP_PREFIX) + ip + ":" + strconv.Itoa(port)
}
